### PR TITLE
Kwxm/lexer tests

### PR DESCRIPTION
**Pull request by kwxm**
This finally gets SCP-430 out of the way.  The lexer now contains some moderately complex stuff for parsing literal constants, so this PR adds some tests to make sure that it can actually deal with inputs that it's supposed to, and that printing followed by parsing is the identity function on constant expressions.  There's some overlap with existing code: I've added a `genConstantForLexerTest` function in `Spec.hs`, but there's already a `genConstant` function in `Generators.AST` which does the same thing but for a more restricted collection of types.  Perhaps we should replace that with the new function, but I think we should wait for Roman to return and consult him before doing that.

Note that this is only testing typed PLC expressions.  We use the same lexer for the untyped parser, so these tests should cover that as well.  PIR uses a megaparsec parser which currently won't be able to handle some literal constants that the Happy parser can.  We could maybe re-use the lexer to tokenise the PIR parser's input.

Pre-submit checklist:
- Branch
    - [x] Commit sequence broadly makes sense
    - [x] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested
- If you updated any cabal files or added Haskell packages:
    - [ ] `$(nix-build default.nix -A pkgsLocal.updateMaterialized)` to update the materialized Nix files
   - [ ] Update `hie-*.yaml` files if needed
- If you changed any Haskell files:
   - [x] `$(nix-shell shell.nix --run fix-stylish-haskell)` to fix any formatting issues
- If you changed any Purescript files:
   - [ ] `$(nix-shell shell.nix --run fix-purty)` to fix any formatting issues

Pre-merge checklist:
- [x] Someone approved it
- [x] Commits have useful messages
- [x] Review clarifications made it into the code
- [ ] History is moderately tidy; or going to squash-merge

### CODE REVIEW

File: plutus-core/test/Spec.hs [hunk 1]
```diff
@@ -102,13 +105,80 @@ propFlat = property $ do
     prog <- forAllPretty $ runAstGen genProgram
     Hedgehog.tripping prog Flat.flat Flat.unflat
 
--- Generate a random 'Program', pretty-print it, and parse the pretty-printed
+{-| The lexer contains some quite complex regular expressions for literal
```
Comment (michaelpj) [inline plutus-core/test/Spec.hs:108]:
```suggestion
{- The lexer contains some quite complex regular expressions for literal
```

File: plutus-core/test/Spec.hs [hunk 2]
```diff
@@ -102,13 +105,80 @@ propFlat = property $ do
...
+  The lexer should be able to consume escape sequences in characters and strings;
+  both standard ASCII escape sequences and Unicode ones.  Hedgehog has generators
+  for both of these, but the Unicode one essentially never generates anything readable:
+  all of the output looks like '\857811'.  For this reason we have separate generators
+  for Unicode characters and Latin-1 ones (characters 0-255, including standard
+  ASCII from 0-127); there is also a generator for UTF8-encoded Unicode. -}
+
+-- TODO: Language.PlutusCore.Generators.AST has a 'genConstant' function, but it only
+-- generates integers and UTF8-encoded Unicode strings. Should we replace that with
+-- this function?
```
Comment (michaelpj) [inline plutus-core/test/Spec.hs:151]: Not sure, perhaps @effectfully has an opinion?

**Review by michaelpj — APPROVED** [review]

File: plutus-core/test/Spec.hs (hunk 2 above)
Comment (kwxm) [reply plutus-core/test/Spec.hs:151]: Yes, but I thought he was on holiday.  I could open another issue to return to this once he's back.

Comment (effectfully) [reply plutus-core/test/Spec.hs:151]: > Should we replace that with this function?

Seems reasonable to me.

> Yes, but I thought he was on holiday. I could open another issue to return to this once he's back.

It doesn't take long to answer a question (sometimes with "dunno, I'll think about it once I'm back"), so feel free to ping me.

Comment (kwxm) [reply plutus-core/test/Spec.hs:151]: OK, maybe I'll stick that in `AST.hs` but still keep separate tests for constants to make sure we get good coverage.  I'll think a bit more about what to do for `()` and `bool`; maybe stick them in there with low frequencies.

Comment (kwxm) [reply plutus-core/test/Spec.hs]: I've now done that.

**Comment by kwxm** [comment]
It turns out to be a bad idea to move the new constant-generator into AST.hs.  Now we generate programs that cause the plutus-ir tests to fail, for example `(program 1.0.0 (con unit ()))`.

**Review by raduom — APPROVED** [review]
LGTM

**Pull request merged**